src/qibo/tensorflow/callbacks.py

```diff
@@ -150,25 +151,41 @@ class EntanglementEntropy(PartialTrace):
             print(entropy[:])
             # Should print [0, 0, 1] which is the entanglement entropy
             # after every gate in the calculation.
+            print(entropy.spectrum)
+            # Print the entanglement spectrum.
     """
     _log2 = tf.cast(tf.math.log(2.0), dtype=DTYPES.get('DTYPE'))
 
+    def __init__(self, partition: Optional[List[int]] = None, compute_spectrum: bool = False):
+        self.compute_spectrum = compute_spectrum
+        self.spectrum = list()
+        super(EntanglementEntropy, self).__init__(partition)
+
     @classmethod
-    def _entropy(cls, rho: tf.Tensor) -> tf.Tensor:
-      """Calculates entropy by diagonalizing the density matrix."""
-      # Diagonalize
-      eigvals = tf.math.real(tf.linalg.eigvalsh(rho))
-      # Treating zero and negative eigenvalues
-      masked_eigvals = tf.gather(eigvals, tf.where(eigvals > EIGVAL_CUTOFF))[:, 0]
-      entropy = - tf.reduce_sum(masked_eigvals * tf.math.log(masked_eigvals))
-      return entropy / cls._log2
+    def _entropy(cls, rho: tf.Tensor, compute_eigvals: bool = False):

```

Comment:
I do not remember exactly why this function is a `@classmethod` and not a normal method, but I tried removing `@classmethod` and tests pass (except two callback tests but these can be fixed). 
I believe that if we make this a normal method we can save a few lines of code for the spectrum as we would be able to do
```Python
spectrum = - tf.math.log(masked_eigvals)
if self.compute_eigvals:
    self.spectrum.append(spectrum)
entropy = tf.reduce_sum(masked_eigvals * spectrum)
return entropy / self._log2
```
directly here and avoid return two values and having the `compute_eigvals` argument.

This is just a suggestion, it is okay as it is now, let me know what you prefer.

Reply:
Yes, I fully agree, I would prefer to remove the `@classmethod` and do everything in one go.
I will propose the changes here.